ODIN_II/SRC/partial_map.cpp

```diff
@@ -903,128 +903,180 @@ void instantiate_GE(nnode_t* node, operation_list type, short mark, netlist_t* n
  *	Creates the hardware for a shift left or right operation by a constant size.
  *-------------------------------------------------------------------------------------------*/
 void instantiate_shift_left_or_right(nnode_t* node, operation_list type, short mark, netlist_t* netlist) {
-    /* these variables are used in an attempt so that I don't need if cases.  Probably a bad idea, but fun */
-    int width;
-    int i;
-    int shift_size;
-    nnode_t* buf_node;
+    
+    if ( node->related_ast_node->children[1]->type != NUMBERS ) {
 
-    width = node->input_port_sizes[0];
+        /*  Create mux 2:1
+        *   data1 = SHIFT first_input
+        *   data2 = SHIFT first_input shifted right or left by pow(2,i)
+        *   selector = SHIFT second_input[i]
+        */ 
 
-    if (node->related_ast_node->children[1]->type == NUMBERS) {
-        /* record the size of the shift */
-        shift_size = node->related_ast_node->children[1]->types.vnumber->get_value();
-    } else {
-        shift_size = 0;
-        error_message(NETLIST, node->loc, "%s\n", "Odin only supports constant shifts at present");
-    }
+        nnode_t** muxes;
+        npin_t** input_pins;
+        npin_t** output_pins;
+        int in_out_port_width;
 
-    buf_node = make_1port_gate(BUF_NODE, width, width, node, mark);
+        input_pins = node->input_pins;
+        output_pins = input_pins;
+        in_out_port_width = node->input_port_sizes[0];
+        muxes = (nnode_t**)vtr::malloc(sizeof(nnode_t*) * (node->input_port_sizes[1]));
```

Comment:
you shouldn't match the input size, at best, it should match the output since thats all that is usable. if you do input, it will blow up in size

Reply:
The **input_port_size[1]** contains the minimum number of bits we should have for the right operand, i.e. **log2(op->children[1]->types.vnumber->get_value())**. However, the **output_port_size** equals to addition of this number with the size of the first shift operand. So, **input_port_sizes[1]** definitely lesser than **output_port_size**. This is also part of Barrel Shifter implementation, so we need use this number of second shift operand input bits to connect to muxes of the next stages.

Reply:
yes and no, as we discussed, all you care about is the bits that can fit in the output. the rest is garbage. 
also YOU CANNOT USE log2(op->children[1]->types.vnumber->get_value()) IN THE NETLIST!

Reply:
to be clear.
```
input [31:0] a;
input [4:0] b;
output [3:0] c;

assign c = a << b;
```

you only need to use b[2:0] for the mux selector. (since c has 4 bit, you need 3 bit to know whether or not your shift operation completely overflowed, and NOR all the other bits above so b[4] and b[3] to make sure they are all 0 otherwise its an overflow)

Reply:
Correction:
I meant **input_port_size[1]** contains the value of **op->children[1]->types.vnumber->size()** which is exactly the bit numbers of the second shift operand. You are right about what we discussed and this is the same thing. This section is part of Barrel Shifter's implementation. So, if we use the size of output port we face segfault since it is equal or greater than the bit numbers of the second shift operand.

Reply:
never do get_value() UNLESS IT IS A NUMBER!

Reply:
Is this (**op->children[1]->types.vnumber->size()**) the right way to find out the size of RANGE_REF?

Reply:
.. what if its just
```
a << b
```
? is that a RANGE_REF?

Reply:
https://github.com/CAS-Atlantic/vtr-verilog-to-routing/blob/master/ODIN_II/SRC/netlist_create_from_ast.cpp#L969-L1097

I would look here to see how they handle different kinds (plain variable, range ref and array ref)